weasyl/middleware.py

```diff
@@ -81,45 +83,65 @@ def session_tween_factory(handler, registry):
     """
     A tween that sets a weasyl_session on a request.
     """
+    def callback(request, response):
+        sess_obj = request.weasyl_session
+
+        if isinstance(sess_obj, GuestSession):
+            if sess_obj.create:
+                response.set_cookie('WZL', sess_obj.sessionid, max_age=None,
+                                    secure=request.scheme == 'https', httponly=True)
+        elif sess_obj.save:
+            session = request.pg_connection
+
+            if sess_obj.create:
+                session.add(sess_obj)
+                response.set_cookie('WZL', sess_obj.sessionid, max_age=60 * 60 * 24 * 365,
+                                    secure=request.scheme == 'https', httponly=True)
+            session.flush()
+        elif sess_obj.userid is None and not sess_obj.additional_data:
+            # An old guest session stored in the database; delete it, but don’t break
+            # its former CSRF token for the duration of this browser session.
+            response.set_cookie('__Host-WZLcsrf', sess_obj.csrf_token, max_age=None,
+                                secure=request.scheme == 'https', httponly=True)
+
+            request.pg_connection.delete(sess_obj)
+            request.pg_connection.flush()
+            response.delete_cookie('WZL')
+
+            captureMessage = request.environ.get('raven.captureMessage')
+            if captureMessage is not None:
+                captureMessage("Upgraded an old guest session", level=logging.INFO)
+
     # TODO(hyena): Investigate a pyramid session_factory implementation instead.
     def session_tween(request):
-        cookies_to_clear = set()
-        if 'beaker.session.id' in request.cookies:
-            cookies_to_clear.add('beaker.session.id')
-
-        session = request.pg_connection
         sess_obj = None
-        if 'WZL' in request.cookies:
-            sess_obj = session.query(orm.Session).get(request.cookies['WZL'])
-            if sess_obj is None:
-                # clear an invalid session cookie if nothing ends up trying to create a
-                # new one
-                cookies_to_clear.add('WZL')
+        cookie = request.cookies.get('WZL')
+
+        if cookie is not None:
+            if is_guest_token(cookie):
```

Comment:
Observation: A peculiar effect of this is the ability to arbitrarily pick one's own ``WZL`` token (e.g., ``12345123451234512345``), and still have a subsequent authentication (a CSRF protected action) succeed. Since we're force-changing the ``WZL`` token upon authentication, it doesn't matter, but I just found that ... slightly amusing (but since we're not saving it anywhere, it makes sense that we have to prime the CSRF token from somewhere).

Reply:
Yeah, it’s a bit awkward and why there’s an `isalnum` check in `is_guest_token` (to prevent self-script-injection from the `'<input type="hidden" name="csrf" value="%s" />' % (user_provided_cookie,)`, which probably doesn’t do anything but has no reason to be allowed). Could be avoided with an HMAC, but that would require more characters; I’m not sure if it’s worth it.